ui/StatusQ/src/safutils_android.cpp

```diff
@@ -0,0 +1,189 @@
+        return -1;
+
+    const auto jUri = jString(documentUri);
+
+    // Signature: static int openWritableFd(String)
+    jint fd = -1;
+    QJniObject ctx = getAndroidContext();
+    // openWritableFd throws in Java; wrap with try-catch by using exceptionCheck
+    fd = QJniObject::callStaticMethod<jint>(
+        "app/status/mobile/SafHelper",
+        "openWritableFd",
+        "(Landroid/content/Context;Ljava/lang/String;)I",
+        ctx.object(), jUri.object<jstring>()
+    );
+
+    // If an exception occurred, Qt will print it in debug; return -1
+    return static_cast<int>(fd);
+}
+
+#include <cstdlib>
+
+extern "C" Q_DECL_EXPORT const char* statusq_saf_copyFromPathToTree(const char* srcPath,
+                                                                     const char* treeUri,
+                                                                     const char* mime,
+                                                                     const char* displayName)
+{
+    if (!srcPath || !*srcPath || !treeUri || !*treeUri)
+        return nullptr;
+
+    const auto jSrc = jString(srcPath);
+    const auto jTree = jString(treeUri);
+    const auto jMime = jString(mime ? mime : "application/octet-stream");
+    const auto jName = jString(displayName ? displayName : "backup.bkp");
+
+    QJniObject ctx = getAndroidContext();
+    QJniObject jResult = QJniObject::callStaticObjectMethod(
+        "app/status/mobile/SafHelper",
+        "copyFromPathToTree",
+        "(Landroid/content/Context;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;)Ljava/lang/String;",
+        ctx.object(), jSrc.object<jstring>(), jTree.object<jstring>(), jMime.object<jstring>(), jName.object<jstring>()
+    );
+
+    const QString uri = jResult.isValid() ? jResult.toString() : QString();
+    if (uri.isEmpty())
+        return nullptr;
+
+    QByteArray utf8 = uri.toUtf8();
+    char* out = static_cast<char*>(std::malloc(static_cast<size_t>(utf8.size()) + 1));
+    if (!out)
+        return nullptr;
+    std::memcpy(out, utf8.constData(), static_cast<size_t>(utf8.size()));
+    out[utf8.size()] = '\0';
+    return out;
+}
+
+extern "C" Q_DECL_EXPORT const char* statusq_saf_getReadableTreePath(const char* treeUri)
+{
+    if (!treeUri || !*treeUri)
+        return nullptr;
+    const auto jTree = jString(treeUri);
+    QJniObject jResult = QJniObject::callStaticObjectMethod(
+        "app/status/mobile/SafHelper",
+        "getReadableTreePath",
+        "(Ljava/lang/String;)Ljava/lang/String;",
+        jTree.object<jstring>()
+    );
+    const QString s = jResult.isValid() ? jResult.toString() : QString();
+    if (s.isEmpty())
+        return nullptr;
+    QByteArray utf8 = s.toUtf8();
+    char* out = static_cast<char*>(std::malloc(static_cast<size_t>(utf8.size()) + 1));
+    if (!out)
+        return nullptr;
+    std::memcpy(out, utf8.constData(), static_cast<size_t>(utf8.size()));
+    out[utf8.size()] = '\0';
```

Comment:
A more convenient way is sth like:
```cpp
 QString tmp = "test";
 QByteArray text = tmp.toUtf8();
 char *data = new char[text.size() + 1];
 strcpy(data, text.data());
 //delete [] data;
```

Reply:
Cool it works! What do you mean by "needs to be freed with `delete [] data;`" though? In the calling function, I need to free the returned value once I'm done with it?

Reply:
Nope, on the usage side, see below